### PR TITLE
Begin moving date_histogram to offset rounding

### PR DESCRIPTION
**Pull request by nik9000**
We added a new rounding in #50609 that handles offsets to the start and
end of the rounding so that we could support `offset` in the `composite`
aggregation. This starts moving `date_histogram` to that new offset.

### CODE REVIEW

**Comment by nik9000** [comment]
Neat! Real test failure.

**Comment by nik9000** [comment]
@elasticmachine, run elasticsearch-ci/2

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ExtendedBounds.java [hunk 1]
```diff
@@ -166,7 +166,11 @@ ExtendedBounds parseAndValidate(String aggName, QueryShardContext queryShardCont
     }
 
     ExtendedBounds round(Rounding rounding) {
-        return new ExtendedBounds(min != null ? rounding.round(min) : null, max != null ? rounding.round(max) : null);
+        // Extended bounds shouldn't be effected by the offset
+        Rounding effectiveRounding = rounding.withoutOffset();
+        return new ExtendedBounds(
+                min != null ? effectiveRounding.round(min) : null,
+                max != null ? effectiveRounding.round(max) : null);
```
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ExtendedBounds.java]: Excellent formatting choice, much more readable.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateRangeHistogramAggregator.java [hunk 2]
```diff
@@ -153,10 +151,6 @@ public void collect(int doc, long bucket) throws IOException {
         };
     }
 
-    private long offsetAwareRounding(Rounding rounding, long value, long offset) {
```
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateRangeHistogramAggregator.java]: Glad to see this getting moved into the rounding itself.  Makes much more sense.

File: server/src/main/java/org/elasticsearch/common/Rounding.java [hunk 3]
```diff
@@ -164,6 +164,19 @@ public void writeTo(StreamOutput out) throws IOException {
      */
     public abstract long nextRoundingValue(long value);
 
+    /**
+     * How "offset" this rounding is from the traditional "start" of the period.
+     * @deprecated We're in the process of abstracting offset *into* Rounding
+     *             so keep any usage to migratory shims
+     */
+    @Deprecated
+    public abstract long offset();
```
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/common/Rounding.java]: What's the plan to migrate away from this? I'm not sure I understand the benefit to adding a deprecated method now that we intend to remove in the next PR, as opposed to just removing the need for it in this PR.

**Review by not-napoleon — CHANGES_REQUESTED** [review]
Looks good overall, but I'm a little concerned about the new & deprecated `offset()` method, as noted.  If there's a plan for that, then I'm ok to merge as is, but I'd like to be explicit about the plan.

File: server/src/main/java/org/elasticsearch/common/Rounding.java (hunk 3 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/common/Rounding.java]: I added this to cut the PR in half, basically. I'll remove it in a followup. I believe that follow up with have serialization changes so it'll *feel* pretty different than this PR.

**Review by not-napoleon — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/common/Rounding.java (hunk 3 above)
Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/common/Rounding.java]: Seems reasonable, thanks for explaining!

**Review by not-napoleon — APPROVED** [review]

**Comment by nik9000** [comment]
Thanks @not-napoleon!

**Comment by nik9000** [comment]
I'm reverting this because it was subtly broken in ways we didn't have tests for. I'll make a PR that adds the tests and redoes it soon enough.

**Pull request merged**